chain/store/name.go

```diff
@@ -117,30 +228,131 @@ func (cs *ChainStore) GetName(registrant []byte) (string, error) {
-				if v, ok := sdb.nameRegistrants.Load(nameId); ok {
-					if registrant, ok := v.([]byte); ok {
-						sdb.updateName(registrant, name)
-						if commit {
-							sdb.nameRegistrants.Delete(nameId)
-						}
-					}
-				}
+		if nameId, ok := key.(string); ok {
+			if info, ok := value.(*nameInfo); ok && !info.Empty() {
+				sdb.updateName(nameId, info)
 			} else {
-				sdb.deleteName(registrantId)
+				sdb.deleteName(nameId)
 			}
 			if commit {
-				sdb.names.Delete(registrantId)
+				sdb.names.Delete(nameId)
 			}
 		}
 		return true
 	})
+
+	sdb.namesCleanup.Range(func(key, value interface{}) bool {
+		if height, ok := key.(uint32); ok {
+			if nc, ok := value.(namesCleanup); ok && len(nc) > 0 {
+				sdb.updateNamesCleanup(height, nc)
+			} else {
+				sdb.deleteNamesCleanup(height)
+			}
+			if commit {
+				sdb.namesCleanup.Delete(height)
+			}
+		}
+		return true
+	})
+}
+
+func (sdb *StateDB) getNamesCleanup(height uint32) (namesCleanup, error) {
+	if v, ok := sdb.namesCleanup.Load(height); ok {
+		if nc, ok := v.(namesCleanup); ok {
+			return nc, nil
+		}
+	}
+
+	enc, err := sdb.trie.TryGet(append(NameCleanupPrefix, getNamesCleanupId(height)...))
+	if err != nil {
+		return nil, fmt.Errorf("[getNamesCleanup]can not get name cleanup from trie: %v", err)
+	}
+
+	nc := make(namesCleanup, 0)
+
+	if len(enc) > 0 {
+		buff := bytes.NewBuffer(enc)
+		ncLength, err := serialization.ReadVarUint(buff, 0)
+		if err != nil {
+			return nil, fmt.Errorf("[getNamesCleanup]Failed to decode state object for name cleanup: %v", err)
+		}
+		for i := uint64(0); i < ncLength; i++ {
+			id, err := serialization.ReadVarString(buff)
+			if err != nil {
+				return nil, fmt.Errorf("[getNamesCleanup]Failed to decode state object for name cleanup: %v", err)
+			}
+			nc[id] = struct{}{}
+		}
+	}
+
+	sdb.namesCleanup.Store(height, nc)
+	return nc, nil
+}
+
+func getNamesCleanupId(height uint32) []byte {
+	buff := bytes.NewBuffer(nil)
+	serialization.WriteUint32(buff, height)
+	return buff.Bytes()
+}
+
+func (sdb *StateDB) CleanupNames(height uint32) error {
+	ids, err := sdb.getNamesCleanup(height)
+	if err != nil {
+		return err
+	}
+	for id := range ids {
+		sdb.names.Store(id, nil)
+	}
+	sdb.namesCleanup.Store(height, nil)
+
+	return nil
+}
+
+func (sdb *StateDB) cleanupNamesAtHeight(height uint32, id []byte) error {
+	ids, err := sdb.getNamesCleanup(height)
+	if err != nil {
+		return err
+	}
+	ids[string(id)] = struct{}{}
+	return nil
+}
+
+func (sdb *StateDB) deleteNamesCleanup(height uint32) error {
+	err := sdb.trie.TryDelete(append(NameCleanupPrefix, getNamesCleanupId(height)...))
+	if err != nil {
+		return err
+	}
+
+	sdb.namesCleanup.Delete(height)
+	return nil
+}
+
+func (sdb *StateDB) updateNamesCleanup(height uint32, nc namesCleanup) error {
+	buff := bytes.NewBuffer(nil)
+
+	if err := serialization.WriteVarUint(buff, uint64(len(nc))); err != nil {
+		panic(fmt.Errorf("can't encode names cleanup %v: %v", nc, err))
+	}
+	ncs := make([]string, 0)
```

Comment:
`make(..., len(nc))` for better efficient (avoid capacity reallocate)
Or
`reflect.ValueOf(nc).MapKeys()` for better concise

Reply:
done